tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_on_block.py

```diff
@@ -703,3 +703,78 @@ def test_new_finalized_slot_is_justified_checkpoint_ancestor(spec, state):
+    yield 'anchor_state', state
+    yield 'anchor_block', anchor_block
+
+    # Build block that serves as head ONLY on timely arrival, and ONLY in that slot
+    state = genesis_state.copy()
+    next_slots(spec, state, 3)
+    block = build_empty_block_for_next_slot(spec, state)
+    signed_block = state_transition_and_sign_block(spec, state, block)
+
+    # Process block on timely arrival just before end of boost interval
+    spec.on_tick(store, store.genesis_time + block.slot * spec.config.SECONDS_PER_SLOT +
+                 spec.config.SECONDS_PER_SLOT // spec.INTERVALS_PER_SLOT - 1)
+    yield from tick_and_add_block(spec, store, signed_block, test_steps)
+    assert store.proposer_boost_root == spec.hash_tree_root(block)
+    assert spec.get_latest_attesting_balance(store, spec.hash_tree_root(block)) > 0
+
+    # Ensure that boost is removed after slot is over
+    spec.on_tick(store, store.genesis_time + block.slot * spec.config.SECONDS_PER_SLOT +
+                 spec.config.SECONDS_PER_SLOT)
+    assert store.proposer_boost_root == spec.Root()
+    assert spec.get_latest_attesting_balance(store, spec.hash_tree_root(block)) == 0
+
+    next_slots(spec, state, 3)
+    block = build_empty_block_for_next_slot(spec, state)
+    signed_block = state_transition_and_sign_block(spec, state, block)
+
+    # Process block on timely arrival at start of boost interval
+    spec.on_tick(store, store.genesis_time + block.slot * spec.config.SECONDS_PER_SLOT)
+    yield from tick_and_add_block(spec, store, signed_block, test_steps)
+    assert store.proposer_boost_root == spec.hash_tree_root(block)
+    assert spec.get_latest_attesting_balance(store, spec.hash_tree_root(block)) > 0
+
+    # Ensure that boost is removed after slot is over
+    spec.on_tick(store, store.genesis_time + block.slot * spec.config.SECONDS_PER_SLOT +
+                 spec.config.SECONDS_PER_SLOT)
+    assert store.proposer_boost_root == spec.Root()
+    assert spec.get_latest_attesting_balance(store, spec.hash_tree_root(block)) == 0
+
+    yield 'steps', test_steps
+
+
+@with_all_phases
+@spec_state_test
+def test_proposer_boost_root_same_slot_untimely_block(spec, state):
+    test_steps = []
+    genesis_state = state.copy()
+
+    # Initialization
+    store, anchor_block = get_genesis_forkchoice_store_and_block(spec, state)
+    yield 'anchor_state', state
+    yield 'anchor_block', anchor_block
+
+    # Build block that serves as head ONLY on timely arrival, and ONLY in that slot
+    state = genesis_state.copy()
+    next_slots(spec, state, 3)
+    block = build_empty_block_for_next_slot(spec, state)
+    signed_block = state_transition_and_sign_block(spec, state, block)
+
+    # Process block on untimely arrival in the same slot
+    spec.on_tick(store, store.genesis_time + block.slot * spec.config.SECONDS_PER_SLOT +
+                 spec.config.SECONDS_PER_SLOT // spec.INTERVALS_PER_SLOT)
+    yield from tick_and_add_block(spec, store, signed_block, test_steps)
```

Comment:
(2) is a no-op on the tick, right? because block.slot time is already in the past?

I assumed `tick_and_add_block` would tick to the greatest value in store or time_of(block.slot), but I see that it could even send time backwards.

You're right and the above fix looks good

Reply:
> (2) is a no-op on the tick, right? because block.slot time is already in the past?

Yes! But using `add_block` would be more accurate in this case.

Reply:
> I assumed `tick_and_add_block` would tick to the greatest value in store or time_of(block.slot), but I see that it could even send time backwards.

`tick_and_add_block` set ticks wiith `max(block_time, store.time)`, where `block_time` is time_of(block.slot)

When I reviewed it today, I was a bit surprised that we didn't check if the `on_tick` input `time` is greater than or equal to `store.time`. Should we add some fool-proof check in the spec?